File: ios/Jazzify/BalloonRush/Engine/BalloonRushSpawn.swift
import CoreGraphics
import Foundation

/// Web `balloonRushSpawn.ts` と同期。
enum BalloonRushSpawn {
    static let initialNearPlayerMinPx: CGFloat = 60
    static let initialNearPlayerMaxPx: CGFloat = 100
    static let initialFarFromPlayerMinPx: CGFloat = 140
    static let betweenBalloonsMinPx: CGFloat = 100
    static let respawnFromPlayerMinPx: CGFloat = 120
    static let mapMarginPx: CGFloat = 48

    struct Point: Sendable {
        var x: CGFloat
        var y: CGFloat
    }

    static func hypotDist(_ a: Point, _ b: Point) -> CGFloat {
        hypot(a.x - b.x, a.y - b.y)
    }

    private static func clampToMap(x: CGFloat, y: CGFloat, margin: CGFloat) -> Point {
        Point(
            x: min(BalloonRushMap.width - margin, max(margin, x)),
            y: min(BalloonRushMap.height - margin, max(margin, y))
        )
    }

    private static func randomInRing(
        cx: CGFloat,
        cy: CGFloat,
        minR: CGFloat,
        maxR: CGFloat,
        margin: CGFloat,
        rng: () -> CGFloat
    ) -> Point {
        let r = minR + rng() * (maxR - minR)
        let ang = rng() * .pi * 2
        return clampToMap(x: cx + cos(ang) * r, y: cy + sin(ang) * r, margin: margin)
    }

    static func pickInitialFivePositions(player: Point, margin: CGFloat, rng: () -> CGFloat) -> [Point] {
        var out: [Point] = []
        let first = randomInRing(
            cx: player.x,
            cy: player.y,
            minR: initialNearPlayerMinPx,
            maxR: initialNearPlayerMaxPx,
            margin: margin,
            rng: rng
        )
        out.append(first)

        for _ in 0..<4 {
            var cand = first
            for _ in 0..<60 {
                let sx = margin + rng() * (BalloonRushMap.width - margin * 2)
                let sy = margin + rng() * (BalloonRushMap.height - margin * 2)
                let trial = clampToMap(x: sx, y: sy, margin: margin)
                if hypotDist(player, trial) < initialFarFromPlayerMinPx { continue }
                var ok = true
                for o in out where hypotDist(trial, o) < betweenBalloonsMinPx {
                    ok = false
                    break
                }
                if ok {
                    cand = trial
                    break
                }
            }
            out.append(cand)
        }
        return out
    }

    static func pickRespawnPosition(
        player: Point,
        existing: [Point],
        margin: CGFloat,
        rng: () -> CGFloat
    ) -> Point? {
        for _ in 0..<120 {
            let sx = margin + rng() * (BalloonRushMap.width - margin * 2)
            let sy = margin + rng() * (BalloonRushMap.height - margin * 2)
            let trial = clampToMap(x: sx, y: sy, margin: margin)
            if hypotDist(player, trial) < respawnFromPlayerMinPx { continue }
            var ok = true
            for o in existing where hypotDist(trial, o) < betweenBalloonsMinPx {
                ok = false
                break
            }
            if ok { return trial }
        }
        return nil
    }
}

File: src/utils/balloonRushSpawn.ts
/**
 * Balloon Rush spawn positions — keep in sync with iOS BalloonRushSpawn.swift
 */
import { BALLOON_RUSH_MAP_CONFIG } from '@/utils/balloonRushMap';

/** ステージ開始: キャラクター近傍の最初の風船 */
export const INITIAL_NEAR_PLAYER_MIN_PX = 60;
export const INITIAL_NEAR_PLAYER_MAX_PX = 100;
/** その他はプレイヤーからこれより遠く */
export const INITIAL_FAR_FROM_PLAYER_MIN_PX = 140;
/** 風船同士・再出現の最小距離 */
export const BETWEEN_BALLOONS_MIN_PX = 100;
/** 再出現時: プレイヤーとの最小距離 */
export const RESPAWN_FROM_PLAYER_MIN_PX = 120;
export const MAP_MARGIN_PX = 48;

export type RandomFn = () => number;

export interface XY {
  readonly x: number;
  readonly y: number;
}

export const hypotDist = (a: XY, b: XY): number => Math.hypot(a.x - b.x, a.y - b.y);

const clampToMap = (x: number, y: number, margin: number): XY => ({
  x: Math.min(BALLOON_RUSH_MAP_CONFIG.width - margin, Math.max(margin, x)),
  y: Math.min(BALLOON_RUSH_MAP_CONFIG.height - margin, Math.max(margin, y)),
});

const randomInRing = (
  cx: number,
  cy: number,
  minR: number,
  maxR: number,
  margin: number,
  rng: RandomFn,
): XY => {
  const r = minR + rng() * (maxR - minR);
  const ang = rng() * Math.PI * 2;
  const x = cx + Math.cos(ang) * r;
  const y = cy + Math.sin(ang) * r;
  return clampToMap(x, y, margin);
};

/** すべての拘束を満たすまで試行（失敗時はラスト候補） */
export const pickInitialFivePositions = (
  player: XY,
  margin: number,
  rng: RandomFn,
): readonly XY[] => {
  const out: XY[] = [];
  const first = randomInRing(
    player.x,
    player.y,
    INITIAL_NEAR_PLAYER_MIN_PX,
    INITIAL_NEAR_PLAYER_MAX_PX,
    margin,
    rng,
  );
  out.push(first);

  const maxAttempts = 60;
  for (let bi = 0; bi < 4; bi += 1) {
    let cand: XY = first;
    for (let a = 0; a < maxAttempts; a += 1) {
      const sx = margin + rng() * (BALLOON_RUSH_MAP_CONFIG.width - margin * 2);
      const sy = margin + rng() * (BALLOON_RUSH_MAP_CONFIG.height - margin * 2);
      const trial = clampToMap(sx, sy, margin);
      const dfp = hypotDist(player, trial);
      if (dfp < INITIAL_FAR_FROM_PLAYER_MIN_PX) {
        continue;
      }
      let ok = true;
      for (const o of out) {
        if (hypotDist(trial, o) < BETWEEN_BALLOONS_MIN_PX) {
          ok = false;
          break;
        }
      }
      if (!ok) {
        continue;
      }
      cand = trial;
      break;
    }
    out.push(cand);
  }

  return out;
};

export const pickRespawnPosition = (
  player: XY,
  existing: readonly XY[],
  margin: number,
  rng: RandomFn,
): XY | null => {
  const maxAttempts = 120;
  for (let a = 0; a < maxAttempts; a += 1) {
    const sx = margin + rng() * (BALLOON_RUSH_MAP_CONFIG.width - margin * 2);
    const sy = margin + rng() * (BALLOON_RUSH_MAP_CONFIG.height - margin * 2);
    const trial = clampToMap(sx, sy, margin);
    if (hypotDist(player, trial) < RESPAWN_FROM_PLAYER_MIN_PX) {
      continue;
    }
    let ok = true;
    for (const o of existing) {
      if (hypotDist(trial, o) < BETWEEN_BALLOONS_MIN_PX) {
        ok = false;
        break;
      }
    }
    if (ok) return trial;
  }
  return null;
};
